porch/controllers/rootsyncrollout/pkg/controllers/rootsyncrollout/controller.go

```diff
@@ -0,0 +1,317 @@
+			controllerutil.RemoveFinalizer(&rootsyncrollout, myFinalizerName)
+			if err := r.Update(ctx, &rootsyncrollout); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update %s after delete finalizer: %w", req.Name, err)
+			}
+		}
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	// Get all packages targeted by the Package selector
+	packageMap, err := r.getPackages(ctx, &rootsyncrollout)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	klog.Infof("Found %d packages: %v", len(packageMap), packageMap)
+
+	clusterList, err := r.getClusters(ctx, &rootsyncrollout)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	klog.Infof("Found %d clusters: %v", len(clusterList), clusterList)
+
+	for pkg := range packageMap {
+		revisions := packageMap[pkg]
+		var latest *porchapi.PackageRevision
+		for i := range revisions {
+			if isLatest(revisions[i]) {
+				l := *revisions[i]
+				latest = &l
+			}
+		}
+
+		for _, cluster := range clusterList {
+			klog.Infof("Checking for cluster %s and packagerevision %s (package %s)", cluster.GetName(), latest.Name, latest.Spec.PackageName)
+			repo, err := r.getRepository(ctx, latest.Spec.RepositoryName, latest.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			rootSyncSetName := fmt.Sprintf("%s-%s", latest.Spec.PackageName, cluster.GetName())
+			rootSyncSetNamespace := req.Namespace
+
+			var rootsyncset rss.RootSyncSet
+			err = r.Get(ctx, types.NamespacedName{
+				Name:      rootSyncSetName,
+				Namespace: rootSyncSetNamespace,
+			}, &rootsyncset)
+
+			if err != nil && !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, err
+			}
+
+			if apierrors.IsNotFound(err) {
+				klog.Infof("Creating rootsyncset %s", rootSyncSetName)
+				newrootsyncset := newRootSyncSet(&rootsyncrollout, cluster, latest, repo)
+				if err := r.Create(ctx, newrootsyncset); err != nil {
+					return ctrl.Result{}, err
+				}
+			} else {
+				updatedrootsyncset := rootsyncset.DeepCopy()
+				updatedrootsyncset.Spec.Template.Spec.Git = &rss.GitInfo{
+					Repo:     repo.Spec.Git.Repo,
+					Revision: fmt.Sprintf("%s/%s", latest.Spec.PackageName, latest.Spec.Revision),
+					Dir:      latest.Spec.PackageName,
+					Branch:   repo.Spec.Git.Branch,
+					Auth:     "none",
+				}
+				if apiequality.Semantic.DeepEqual(updatedrootsyncset, &rootsyncset) {
+					klog.Infof("No changes to rootsyncset %s", rootsyncset.Name)
+					continue
+				}
+				klog.Info("Updating rootsyncset %s", rootsyncset.Name)
+				if err := r.Update(ctx, updatedrootsyncset); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+		}
+
+	}
+
+	nextReconcile, found := r.NextScheduledReconcile[req.NamespacedName]
+	if !found {
+		nextReconcile = time.Time{}
+	}
+	if nextReconcile.After(time.Now().Add(2 * time.Second)) {
+		return ctrl.Result{}, nil
+	}
+
+	newReconcile := time.Now().Add(30 * time.Second)
+	r.NextScheduledReconcile[req.NamespacedName] = newReconcile
```

Comment:
Do we need to cache this information?  I've always seen it just returning RequeueAfter...

Reply:
Not sure I follow. Are you saying there is a way to query the workqueue to check when the next reconcile is scheduled for? This solution is just temporary until we have support for watches on packagerevisions and I've figured out how to set up watches correctly with labelselectors. That is the next steps here.